pkgs/misc/fastly/default.nix

```diff
@@ -0,0 +1,42 @@
+{ lib, fetchFromGitHub, go, installShellFiles, buildGoModule }:
+
+buildGoModule rec {
+  pname = "fastly";
+  version = "3.1.0";
+
+  src = fetchFromGitHub {
+    owner = "fastly";
+    repo = "cli";
+    rev = "v${version}";
+    sha256 = "sha256-ZztaJM87O7T8dl4NMQqVHBqJPdSf18hrh4tYJissVkM=";
+  };
+
+  subPackages = [ "cmd/fastly" ];
+
+  vendorSha256 = "sha256-5MvJS10f7YLvO+wCmUJleU27hCJbsNrOIfUZnniGw+E=";
+
+  nativeBuildInputs = [ installShellFiles ];
+
+  # Flags as provided by the build automation of the project:
+  #   https://github.com/fastly/cli/blob/7844f9f54d56f8326962112b5534e5c40e91bf09/.goreleaser.yml#L14-L18
+  ldflags = ["-s" "-w"] ++ (with lib;
+    mapAttrsToList (n: v: ''-X "github.com/fastly/cli/pkg/revision.${n}=${v}"'') {
+      AppVersion = "v${version}";
+      GitCommit = builtins.substring 0 7 src.rev;
+      GoVersion = lib.getVersion go;
```

Comment:
you've pretty much covered everything I was going to comment on while I was working on an upstream patch. Here are the fixes:

---

You'll want to use `buildGoModule.go` to use the same copy. (this isn't necessary for the more dirty `preBuild` method at the bottom)

---

I've ran into this issue of providing host build info a couple times (https://github.com/NixOS/nixpkgs/pull/177868), the go version should be pulled in by `runtime.Version()` really (I opened a PR https://github.com/fastly/cli/pull/588) but we will still need to give the host OS and ARCH once that's merged

you can get these from `buildGoModule.go.GOHOSTOS` & `buildGoModule.go.GOHOSTARCH` either by grabbing them like that

or you could inherit the values somewhere and refer to them by shorthand
`inherit (buildGoModule.go) GOHOSTOS GOHOSTARCH;`

(unlike kdigger which was relying on `uname -m` we really want to get the values via `buildGoModule.go` because there's been some modification to convert things e.g. `aarch64` -> `arm64`)

If you don't want to adopt the patch ahead of it being at least reviewed or merged then below are some more options related to the sha

---

you can either leave the commit sha and set it to `unknown` and be done with it or you can use this method:

Get the info and keep it in the src (we make sure to delete the .git folder after since it causes issues)

**You must update the src sha or you'll just be using the old cached src in your nix store**
```
# populate values that require us to use git. By doing this in postFetch we
# can delete .git afterwards and maintain better reproducibility of the src.
leaveDotGit = true;
postFetch = ''
  cd "$out"
  git rev-parse HEAD > $out/COMMIT
  find "$out" -name .git -print0 | xargs -0 rm -rf
'';
```

and then we add to the original ldflags in the build, this is also where we could just use `go version` directly for now until the patch is merged and it can be done in a more nix friendly way
```
# ldflags based on metadata from git and source
preBuild = ''
  ldflags+=" -X github.com/fastly/cli/pkg/revision.GitCommit=$(cat COMMIT)"
  ldflags+=" -X github.com/fastly/cli/pkg/revision.GoVersion=$(go version)"
'';
```

Reply:
I think it would be desirable that the derivation is able to refer to the "go used to build it". If that changes on the caller (something like what you posted in all-packages), it's transparent to the fastly derivation.

I see that zabbbix agent2 has something similar, but is only able to access `GOOS` and `GOARCH`.

Reply:
trickster should just be grabbing GOARCH from the `runtime` package since it looks like the target's arch rather than the builder's arch

Reply:
Thanks for the pointers @06kellyjac. I just updated the PR. However the `go` reference is still wrong, and `buildGoModule.go` attribute does not exist.